src/librustc_mir/transform/promote_consts.rs

```diff
@@ -134,22 +157,644 @@ impl<'tcx> Visitor<'tcx> for TempCollector<'tcx> {
+                            has_mut_interior = false;
+                            break;
+                        }
+
+                        place_projection = proj_base;
+                    }
+                }
+                if has_mut_interior {
+                    return Err(Unpromotable);
+                }
+
+                Ok(())
+            }
+
+            Rvalue::Aggregate(_, ref operands) => {
+                for o in operands {
+                    self.validate_operand(o)?;
+                }
+
+                Ok(())
+            }
+        }
+    }
+
+    fn validate_call(
+        &self,
+        callee: &Operand<'tcx>,
+        args: &[Operand<'tcx>],
+    ) -> Result<(), Unpromotable> {
+        if !self.explicit && self.is_non_const_fn {
+            if let ty::FnDef(def_id, _) = callee.ty(self.body, self.tcx).kind {
+                // Never promote runtime `const fn` calls of
+                // functions without `#[rustc_promotable]`.
+                if !self.tcx.is_promotable_const_fn(def_id) {
+                    return Err(Unpromotable);
+                }
+            }
+        }
+
+        let fn_ty = callee.ty(self.body, self.tcx);
+        match fn_ty.kind {
+            ty::FnDef(def_id, _) => {
+                match self.tcx.fn_sig(def_id).abi() {
+                    Abi::RustIntrinsic |
+                    Abi::PlatformIntrinsic => {
+                        assert!(!self.tcx.is_const_fn(def_id));
+                        match &self.tcx.item_name(def_id).as_str()[..] {
+                            | "size_of"
+                            | "min_align_of"
+                            | "needs_drop"
+                            | "type_id"
+                            | "bswap"
+                            | "bitreverse"
+                            | "ctpop"
+                            | "cttz"
+                            | "cttz_nonzero"
+                            | "ctlz"
+                            | "ctlz_nonzero"
+                            | "wrapping_add"
+                            | "wrapping_sub"
+                            | "wrapping_mul"
+                            | "unchecked_shl"
+                            | "unchecked_shr"
+                            | "rotate_left"
+                            | "rotate_right"
+                            | "add_with_overflow"
+                            | "sub_with_overflow"
+                            | "mul_with_overflow"
+                            | "saturating_add"
+                            | "saturating_sub"
+                            | "transmute"
+                            | "simd_insert"
+                            | "simd_extract"
+                            => return Err(Unpromotable),
+
+                            _ => {}
+                        }
+                    }
+                    _ => {
+                        let is_const_fn =
```

Comment:
We should deny promoting any functions unless they have the `rustc_promotable` attribute

Reply:
Oh it does so further up, so this part is only for inside const fn or explicit constants, carry on